### PR TITLE
fetch: Change patch format

### PR DESCRIPTION
**Pull request by rafasc**
Available formats:
    - log: show all commits between HEAD and last seen.
    - diff: show differences between HEAD and last seen.

### CODE REVIEW

**Comment by rafasc** [comment]
In this PR, the new `diff` format is made default for the sake of testing. That decision is up to discussion. So are the formats  themselves, as both have pros and cons.

Maybe the ideal would be having a combination of these two: `diff` at the top, some kind of separator, followed by `log`.

Depending on feedback, this may be postponed until the next milestone.

Cc: @maximbaz as he helped me test this feature.

**Comment by AladW** [comment]
Could you maybe add some screenshots on the different formats?

**Comment by rafasc** [comment]
Consider you run fetch, and there are 3 new commits. 

The `diff` format will generate a single diff considering the before and after, while `log` will show the current `git log` based patch, which shows 1 diff per commit, and its respective message.

I am not sure if a screenshot is the best way to show the difference, but here's an attempt:
formats

`diff` results in a much shorter patch, making the review easier at the expense of not showing intermediate changes.

**Comment by rafasc** [comment]
As it is impossible to cover everyone's cases, another interesting idea would be to allow a format `git-<name>`; where `gen_patchfile` would call `$git name;`, allowing users to define `git` aliases that control the format.

We could support a couple of sane defaults, and leave the rest for the user to implement as an alias.

Or go the other route, and use a predetermined `git alias` that if found would be used, falling back to what we do currently.

**Comment by AladW** [comment]
From a purely personal perspective, I prefer the `diff` format since I'm usually not very interested in scrolling the page to view all commits. Especially when every commit includes changes to .SRCINFO as well...

I guess a `git alias` could be placed in another environment variable, though I'm unsure if we need this additional granularity. It's hard to compete to dedicated GUIs like `tig` anyway, e.g.

tig

**Comment by rafasc** [comment]
Since I also prefer `diff` myself, maybe we can simplify the logic and use a predetermined alias if it exists, falling back to `diff` mode. I.e.:

``` sh
if git config --get alias.aurutils-patch-gen; then
    git aurutils-patch-gen "${log_range[@]}"
else
    git diff ... "${log_range[@]}"
fi
```
The current mode (`log`), could be achieved via the `git alias` mechanism if desired.

The user can configure the `git alias` globally in their `~/.gitconfig` , or per repo if they want. And since git aliases can be shell functions, that gives it a lot of flexibility and removes our burden to support an endless list of formats to suit everyone's needs.

`vifm` can be made to launch `tig` automatically with "`:autocmd DirEnter .git !tig`" in `vifmrc` but, unfortunately, doing it destroys the `vifm view` we create.

For those reasons, I think there's value in supporting the `git alias` feature.

**Comment by maximbaz** [comment]
I also prefer `diff`, maybe we should just switch to `diff` entirely and wait with further generalizations until we actually hear from someone not happy that `log` got replaced with `diff`?

**Comment by rafasc** [comment]
I was trying to not lose the feature of being able to conveniently read the log.

But since `vifm` can be made to display repository entries in a friendlier way with:
```
autocmd DirEnter ~/.cache/aurutils/sync/*/.git,/tmp/view.????????/*/.git !tig
fileviewer {{~/.cache/aurutils/sync/*/.git/,/tmp/view.????????/*/.git/}}
           \ git -C %c log --color HEAD $(git -C %c rev-parse --revs-only ^AURUTILS_SEEN)
```
there is no net-loss from dropping `"log"`.

The above could be added as documentation (after changing some of the hardcoded values with their correspondent environment variables) in a PR that also makes the default view tmpdir a bit more recognizable as coming from aurutils. (not /tmp/view.XXXXXXXX)

Thanks for the feedback.

File: lib/aur-fetch [hunk 1]
```diff
@@ -9,6 +9,10 @@ readonly AUR_COLOR=${AUR_COLOR-auto}
 # default options
 verbose=0 recurse=0 fetch_args=('--verbose') confirm_seen=0
 
+gen_patchfile() {
+    git diff-tree --patch --stat "${2:-$(git hash-object -t tree /dev/null)}" "$1";
```
Comment (AladW) [inline lib/aur-fetch]: Nitpick: redundant semi-colon at the end (same for L101/103)

**Comment by AladW** [comment]
:+1:

**Pull request merged**